frame/support/procedural/src/pallet/expand/genesis_build.rs

```diff
@@ -29,33 +29,38 @@ pub fn expand_genesis_build(def: &mut Def) -> proc_macro2::TokenStream {
 
 	let frame_support = &def.frame_support;
 	let type_impl_gen = &def.type_impl_generics(genesis_build.attr_span);
-	let type_use_gen = &def.type_use_generics(genesis_build.attr_span);
-	let trait_use_gen = if def.config.has_instance {
-		quote::quote_spanned!(genesis_build.attr_span => T, I)
-	} else {
-		// `__InherentHiddenInstance` used by construct_runtime here is alias for `()`
-		quote::quote_spanned!(genesis_build.attr_span => T, ())
-	};
 	let gen_cfg_ident = &genesis_config.genesis_config;
 
 	let gen_cfg_use_gen = genesis_config.gen_kind.type_use_gen(genesis_build.attr_span);
 
 	let where_clause = &genesis_build.where_clause;
 
-	quote::quote_spanned!(genesis_build.attr_span =>
-		#[cfg(feature = "std")]
-		impl<#type_impl_gen> #frame_support::sp_runtime::BuildModuleGenesisStorage<#trait_use_gen>
-			for #gen_cfg_ident<#gen_cfg_use_gen> #where_clause
-		{
-			fn build_module_genesis_storage(
-				&self,
-				storage: &mut #frame_support::sp_runtime::Storage,
-			) -> std::result::Result<(), std::string::String> {
-				#frame_support::BasicExternalities::execute_with_storage(storage, || {
-					<Self as #frame_support::traits::GenesisBuild<#type_use_gen>>::build(self);
-					Ok(())
-				})
-			}
+	let assimilate_storage_fn = quote::quote!(
+		fn assimilate_storage(&self, storage: &mut sp_runtime::Storage) -> std::result::Result<(), std::string::String> {
+			#frame_support::BasicExternalities::execute_with_storage(storage, || {
+				self.build();
+				Ok(())
+			})
 		}
-	)
+	);
+
+	if genesis_config.gen_kind.is_generic() {
+		quote::quote_spanned!(genesis_build.attr_span =>
+			#[cfg(feature = "std")]
+			impl<#type_impl_gen> #frame_support::sp_runtime::BuildStorage
+				for #gen_cfg_ident<#gen_cfg_use_gen> #where_clause
+			{
+				#assimilate_storage_fn
+			}
+		)
+	} else {
+		quote::quote_spanned!(genesis_build.attr_span =>
+			#[cfg(feature = "std")]
+			impl #frame_support::sp_runtime::BuildStorage
+				for #gen_cfg_ident
+			{
+				#assimilate_storage_fn
+			}
+		)
+	}
```

Comment:
```suggestion
		quote::quote_spanned!(genesis_build.attr_span =>
			#[cfg(feature = "std")]
			impl<#type_impl_gen> #frame_support::sp_runtime::BuildStorage
				for #gen_cfg_ident<#gen_cfg_use_gen> #where_clause
			{
				fn assimilate_storage(&self, storage: &mut sp_runtime::Storage) -> std::result::Result<(), std::string::String> {
			#frame_support::BasicExternalities::execute_with_storage(storage, || {
				self.build();
				Ok(())
			})
		}
			}
		)
```

You can write `impl<> Test<>` and rust will not complain. So, no need for this `if`.